### PR TITLE
OsuPlayer.Network/IO Unittests

### PR DESCRIPTION
**Pull request by Founntain**
The OsuPlayer.Network Module now has a test coverage of 84%

### CODE REVIEW

File: OsuPlayer.Tests/OsuPlayer.Network/ApiUserEndpointTests.cs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        Assert.IsNull(offlineResult);
+        Assert.IsNull(offlineResult2);
+        Assert.IsNull(offlineResult3);
+        
+        Constants.OfflineMode = false;
+
+        var onlineResult = await ApiAsync.GetRequestAsync<User>("nonExistingController", "nonExistingAction");
+        var onlineResult2 = await ApiAsync.ApiRequestAsync<User>("nonExistingController", "nonExistingAction", "");
+        var onlineResult3 = await ApiAsync.GetRequestWithParameterAsync<User>("nonExistingController", "nonExistingAction", "");
+
```
Comment (Cesan) [inline OsuPlayer.Tests/OsuPlayer.Network/ApiUserEndpointTests.cs:31]: Please split into different test case. Also as mentioned before, all online tests will fail if the api is offline, but shouldn't as we have the explicit offline mode for this case, so we have to discuss practicability of these tests.

File: OsuPlayer.Tests/OsuPlayer.Network/PasswordManagerTests.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    public async Task PasswordManagerTest()
+    {
+        var r1 = PasswordManager.CheckIfPasswordMeetsRequirements("");
+        var r2 = PasswordManager.CheckIfPasswordMeetsRequirements("Test");
+        var r3 = PasswordManager.CheckIfPasswordMeetsRequirements("Test123!_?");
+        
+        Assert.IsFalse(r1);
+        Assert.IsFalse(r2);
+        Assert.IsTrue(r3);
+    }
```
Comment (Cesan) [inline OsuPlayer.Tests/OsuPlayer.Network/PasswordManagerTests.cs:19]: Please remove the async/Task on this test as it does not use any awaits. Also split into different test cases, as already stated earlier.

File: OsuPlayer.Tests/OsuPlayer.IO/PlaylistTests.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+{
+    [Test]
+    public async Task SetCurrentPlaylist()
+    {
+        PlaylistManager.SetCurrentPlaylist(default);
+        PlaylistManager.SetCurrentPlaylist(new ());
+        
+        await PlaylistManager.SetCurrentPlaylistAsync(default);
+        await PlaylistManager.SetCurrentPlaylistAsync(new ());
+    }
```
Comment (Cesan) [inline OsuPlayer.Tests/OsuPlayer.IO/PlaylistTests.cs:21]: Please split synced/async Tests as they are different cases. Does apply for every test case with both calls.

File: OsuPlayer.Tests/OsuPlayer.IO/PlaylistTests.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        var playlists = await PlaylistManager.GetAllPlaylistsAsync();
+        
+        Assert.IsTrue(playlists.Any(x => x.Name == "New Name"));
+    }
+
+    [Test]
+    public async Task SavePlaylist()
+    {
+        await PlaylistManager.SavePlaylistsAsync();
+    }
```
Comment (Cesan) [inline OsuPlayer.Tests/OsuPlayer.IO/PlaylistTests.cs:93]: Synced test missing.

File: OsuPlayer.Tests/OsuPlayer.IO/SongImporterTests.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+{
+    [Test]
+    public async Task ImporterTests()
+    {
+        //We can't run an actual import, because what path should be use obviously.
+        //So we are running negative tests here :(
+        
+        await SongImporter.ImportSongsAsync(string.Empty);
+        await SongImporter.ImportSongsAsync("U:\\HOPEFULLYNOTEXISTINGPATH");
+    }
```
Comment (Cesan) [inline OsuPlayer.Tests/OsuPlayer.IO/SongImporterTests.cs:17]: Split into different test cases either with `[TestCase("xxx")]` or 2 independent tests. Also use Path.Combine as defining "U:\\xxx" statically could potentially cause fails.

**Review by Cesan — CHANGES_REQUESTED** [review]
See comments. Needs work.

File: OsuPlayer.Tests/OsuPlayer.Network/GitHubUpdaterTests.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Threading.Tasks;
+using NUnit.Framework;
+using OsuPlayer.Network;
+
+namespace OsuPlayer.Tests.OsuPlayer.Network;
+
+public class GitHubUpdaterTests
+{
+    [Test]
+    public async Task CheckForUpdatesTest()
```
Comment (Cesan) [inline OsuPlayer.Tests/OsuPlayer.Network/GitHubUpdaterTests.cs:10]: This apparently runs into problems where the github api limits trigger and we can't get a valid response. So likely removing this test completely for now would be best.

**Review by Cesan — CHANGES_REQUESTED** [review]

File: OsuPlayer.Tests/OsuPlayer.IO/PlaylistTests.cs (hunk 4 above)
Comment (Founntain) [reply OsuPlayer.Tests/OsuPlayer.IO/PlaylistTests.cs:93]: Only has an async call.

File: OsuPlayer.Tests/OsuPlayer.Network/GitHubUpdaterTests.cs (hunk 6 above)
Comment (Founntain) [reply OsuPlayer.Tests/OsuPlayer.Network/GitHubUpdaterTests.cs:10]: Yeah makes sense saw that error already. Lets remove them for now

File: OsuPlayer.Tests/OsuPlayer.IO/SongImporterTests.cs (hunk 5 above)
Comment (Founntain) [reply OsuPlayer.Tests/OsuPlayer.IO/SongImporterTests.cs:17]: Splitted them

File: OsuPlayer.Tests/OsuPlayer.IO/PlaylistTests.cs (hunk 3 above)
Comment (Founntain) [reply OsuPlayer.Tests/OsuPlayer.IO/PlaylistTests.cs:21]: Done

File: OsuPlayer.Tests/OsuPlayer.Network/PasswordManagerTests.cs (hunk 2 above)
Comment (Founntain) [reply OsuPlayer.Tests/OsuPlayer.Network/PasswordManagerTests.cs:19]: oops my bad shouldn't be there tbh. Probably a copy paste error

File: OsuPlayer.Tests/OsuPlayer.Network/ApiUserEndpointTests.cs (hunk 1 above)
Comment (Founntain) [reply OsuPlayer.Tests/OsuPlayer.Network/ApiUserEndpointTests.cs:31]: Yeah we can discuss them I splitted them and set them on ignore

File: OsuPlayer.Tests/OsuPlayer.Network/GitHubUpdaterTests.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        var result = await GitHubUpdater.CheckForUpdates();
+        
+        Assert.IsFalse(result.Item1);
```
[the remark this answers is not in the document]
Comment (Founntain) [reply OsuPlayer.Tests/OsuPlayer.Network/GitHubUpdaterTests.cs:14]: As we discussed it down below, I removed them for now

**Review by Cesan — APPROVED** [review]

**Pull request merged**